netkan/netkan/webhooks/github_mirror.py

```diff
@@ -0,0 +1,315 @@
+        "EFL-2.0"           : 'https://opensource.org/licenses/EFL-2.0',
+        "Expat"             : 'https://opensource.org/licenses/MIT',
+        "MIT"               : 'https://opensource.org/licenses/MIT',
+        "GPL-1.0"           : 'http://www.gnu.org/licenses/old-licenses/gpl-1.0.en.html',
+        "GPL-2.0"           : 'http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html',
+        "GPL-3.0"           : 'http://www.gnu.org/licenses/gpl-3.0.en.html',
+        "LGPL-2.0"          : 'https://www.gnu.org/licenses/old-licenses/lgpl-2.0.html',
+        "LGPL-2.1"          : 'https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html',
+        "LGPL-3.0"          : 'http://www.gnu.org/licenses/lgpl-3.0.en.html',
+        "GFDL-1.1"          : 'http://www.gnu.org/licenses/old-licenses/fdl-1.1.en.html',
+        "GFDL-1.2"          : 'http://www.gnu.org/licenses/old-licenses/fdl-1.2.html',
+        "GFDL-1.3"          : 'http://www.gnu.org/licenses/fdl-1.3.en.html',
+        "LPPL-1.0"          : 'https://latex-project.org/lppl/lppl-1-0.html',
+        "LPPL-1.1"          : 'https://latex-project.org/lppl/lppl-1-1.html',
+        "LPPL-1.2"          : 'https://latex-project.org/lppl/lppl-1-2.html',
+        "LPPL-1.3c"         : 'https://latex-project.org/lppl/lppl-1-3c.html',
+        "MPL-1.1"           : 'https://www.mozilla.org/en-US/MPL/1.1/',
+        "Perl"              : 'http://dev.perl.org/licenses/',
+        "Python-2.0"        : 'https://www.python.org/download/releases/2.0/license/',
+        "QPL-1.0"           : 'https://opensource.org/licenses/QPL-1.0',
+        "W3C"               : 'https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document',
+        "Zlib"              : 'http://www.zlib.net/zlib_license.html',
+        "Zope"              : 'http://old.zope.org/Resources/License.1',
+        "Unlicense"         : 'https://unlicense.org/UNLICENSE',
+    }
+
+    def __init__(self, filename):
+        self.filename = Path(filename)
+        self.contents = self.filename.read_text()
+        self._raw     = json.loads(self.contents)
+
+    def can_mirror(self):
+        return (
+            self._raw.get('kind', 'package') == 'package'
+            and self._raw['download_content_type'] in Ckan.mime_to_extension
+            and self.redistributable()
+        )
+
+    def authors(self):
+        auth = self._raw['author']
+        if isinstance(auth, list):
+            return auth
+        else:
+            return [auth]
+
+    def licenses(self):
+        lic = self._raw['license']
+        if isinstance(lic, list):
+            return lic
+        else:
+            return [lic]
+
+    def license_urls(self):
+        return [Ckan.license_urls[lic] for lic in self.licenses() if lic in Ckan.license_urls]
+
+    def redistributable(self):
+        for lic in self.licenses():
+            if lic in Ckan.redistributable_licenses:
+                return True
+        return False
```

Comment:
This doesn't match the current logic. We respect the most restrictive license within the package, unless all components are allowed to be redistributed freely, then we can't upload to the IA on the authors behalf.

Reply:
Hmm, that's a straight port of the current logic:

https://github.com/KSP-CKAN/NetKAN-bot/blob/4f218a2085f9a5e49f151c499a9192e97c5af728/lib/App/KSP_CKAN/Metadata/Ckan.pm#L202-L218

Remember that multiple licenses associated with a module do **not** indicate separate sub-licenses for different pieces, but rather multiple *options* for licensing for the overall package that the user can choose (that's why lists are so rarely used here):

https://github.com/KSP-CKAN/CKAN/blob/master/Spec.md#license

> A single license (v1.0) , or list of licenses (v1.8) may be provided. The following are both valid, the first describing a mod released under the BSD license, the second under the user's choice of BSD-2-clause or GPL-2.0 licenses.

So if there's a redistributable license in a list of multiple, that means that the entire package can be used under that license, not that some piece of it can be.

Reply:
Yes, that's generally how licenses work; most permissive wins. However my understanding is that the metadata can't separate say source code under a permissive license, assets under a restricted license. So in the case of archiving we don't have the rights to redistribute unless the entire mod is covered by a permissive license. See KSP-CKAN/CKAN#22

I'm ok leaving it as the status quo, but technically I think my original logic is wrong.

Reply:
> the metadata can't separate say source code under a permissive license, assets under a restricted license

That's exactly right. A mod in that situation would list only the more restricted license in the `license` property, not both.

Reply:
Oh good, the spec covers that! Well carry on :slightly_smiling_face: